### PR TITLE
FI: Add swapping events

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

- Add `SwapCreated` and `SwapFullfilled` events from the PoV of FI.

**Should be merged before RU**

### CODE REVIEW

File: pallets/foreign-investments/src/impls.rs
Comment (lemunozm) [inline pallets/foreign-investments/src/impls.rs:230]: Note that `..swap.clone()` contains the `currency_in` and `currency_out` to know the swap direction.

**Review by lemunozm — COMMENTED** [review]
@hieronx, I think not, but tell me if you think some information is missing in the events.

File: libs/traits/src/swaps.rs
Comment (wischli) [inline libs/traits/src/swaps.rs:123]: The `OrderInfo` returned by `TokenSwaps::get_order_details` contains a `ratio` field of type `OrderRatio<Ratio>`. Maybe that suffices such that we don't need to introduce an additional ratio field?

**Review by wischli — COMMENTED** [review]
Quick like a spaceship! One comment regarding a potential redundancy. Moreover, it would be nice to assert the newly added events in a unit test.

**Comment by lemunozm** [comment]
> Moreover, it would be nice to assert the newly added events in a unit test.

Agree!  I'll do it in a follow-up PR to not block this in case I have no time to finish them

**Comment by lemunozm** [comment]
Modified to return always the market ratio in a fulfilled event even when it comes from a cancelation.

File: libs/traits/src/swaps.rs
Comment (wischli) [reply libs/traits/src/swaps.rs:123]: Thanks for the explanation! Makes sense to keep as is then.

**Review by wischli — DISMISSED** [review]
LGTM!

**Comment by lemunozm** [comment]
We should not merge it yet. There are some requested changes from slack

File: pallets/foreign-investments/src/lib.rs
Comment (lemunozm) [inline pallets/foreign-investments/src/lib.rs:297]: @hieronx I've needed to add a new `SwapCancelled`. All events are seen from the perspective of who did the swap. 

> Wouldn’t it possible to add an optional arg to the decrease fulfilled event, with the swap ID that was used to cancel it out?

This can not be possible due that swap ID is the same for both directions. Note that swap id is not the order id, which is unknown from FI perspective. As a workaround, I've added the `opposite_in` field to `SwapCancelled,` which has the amount in the opposite direction that has not been canceled. That way, if opposite_in is 0, that means that the cancellation has also canceled the opposite order. That is what you need to know.

Comment (hieronx) [reply pallets/foreign-investments/src/lib.rs:297]: Looks good to me! Only one nit: in `SwapFulfilled`, wouldn't it make more sense to have the `swapped_out` param before the `swapped_in` param?

Comment (lemunozm) [reply pallets/foreign-investments/src/lib.rs:297]: What do you mean by "before" in this context? Placed in swapped line numbers? Probably, it makes more sense. But right now, we have such conventions in all places in the codebase. I'd prefer to keep it as it is everywhere. But we should probably look into it soon 👍🏻

**Comment by wischli** [comment]
> Just RC as I want to see before merge

Just triggered a WASM generation and can update Dev later.

File: pallets/foreign-investments/src/lib.rs
Comment (hieronx) [reply pallets/foreign-investments/src/lib.rs:297]: Well we should never change these event definitions again after merging this, at least not without a very strong reason ;) Would be a pain on the subquery side. But sure it is fine to keep it as is then if that is the convention now

File: pallets/foreign-investments/src/impls.rs
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls.rs:245]: Could it be that both `status.swapped` and `status.pending` are non zero? 

IF YES: 
* How would one interpret the `swap_id` of the two events?

**Review by mustermeiszer — COMMENTED** [review]
Code looks good!

I am still a bit confused to follow the reasoning of the code, but the last time I had the same and found no issues. What would help me if @lemunozm could describe what are the possible states after `apply_swap`, how are they reflected in the `SwapStatus` and the `SwapOf`.

File: pallets/foreign-investments/src/impls.rs
Comment (lemunozm) [reply pallets/foreign-investments/src/impls.rs:245]: Yes, it could. `swap_id` will be the same for both. Note that `swap_id` is just a tuple `(T::InvestmentId, Action)` not an `order_id` (which is a number from `order-book`).

If the above two events are dispatched, it means that when you apply a swap, part of the amount was automatically fulfilled by cancelation with an inverse order (notified by `SwapCancelled` event), and part of the swap could not be fulfilled with the inverse order: such remaining amount needs to be swapped (notified it by `SwapCreated` event).

Possible states after applying swaps are:
- `pending_amount == 0` and `swapped amount > 0` => it means applying a swap has partially or fully canceled any previous inverse swaps. The required amount is obtained from the inverse amount. No more amount needs to be swapped. OrderBook will Not create a new order.
- `pending amount > 0` and `swapped amount > 0` => It means applying a swap has fully canceled any previous inverse swaps, but more amount needs to be swapped. Orderbook will create a new order.
- `pending amount > 0` and `swapped amount == 0` => There was no inverse order, so we need to create a new order.


NOTE: Probably, a better name for `swap_status.swapped` is `swap_status.canceled`. It all depends on how you interpret it.

Comment (mustermeiszer) [reply pallets/foreign-investments/src/impls.rs:245]: Thanks a lot, that helped!

**Review by mustermeiszer — APPROVED** [review]

**Review by wischli — APPROVED** [review]

**Pull request merged**
